### PR TITLE
Feature/run cl command

### PR DESCRIPTION
**Pull request by worksofliam**
### Changes

Adds ability to run CL commands from the active editor. The reason behind this PR is to make it easier to create and maintain objects from local sources, removing the need to compile as a CL program. It also allows access to standard compile time variables.

Kind of solves the problem in #1269.

* Command/Control + R when connected to either selected command, or if nothing is selected, the command at the cursor.
* Shows message with a button to open the spool file output if it is available.

image

### Checklist

* [x] have tested my change
* [ ] updated relevant documentation
* [ ] Remove any/all `console.log`s I added
* [ ] eslint is not complaining
* [ ] have added myself to the contributors' list in [CONTRIBUTING.md](https://github.com/halcyon-tech/vscode-ibmi/blob/master/CONTRIBUTING.md)
* [ ] **for feature PRs**: PR only includes one feature enhancement.

### CODE REVIEW

**Comment by worksofliam** [comment]
Side question: should this really be part of the CL extension?

File: src/languages/clle/clRunner.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  if (selection.isEmpty) {
+    let line = selection.start.line;
+
+    // First let's find out if this command belong to another command
+    if ((line-1) >= 0) {
+      let preLine = document.lineAt(line-1).text.trim();
+
+      while ((line-1) >= 0 && preLine.endsWith(`+`)) {
+        line--;
+        preLine = document.lineAt(line-1).text.trim();
```
Comment (chrjorgensen) [inline src/languages/clle/clRunner.ts:70]:
```suggestion
      while ((line-1) >= 0 && preLine.endsWith(`+`)) {
        preLine = document.lineAt(line-1).text.trim();
        line--;
```

...or `line-1` in line 69 will be -1 at the first line in source, and the `lineAt` will fail and the command not extracted...

File: src/languages/clle/clRunner.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            }
+            
+            const commandResult = await connection!.runCommand({
+              command: selectedCommand.content,
+              environment: `ile`
+            });
+
+            if (commandResult) {
+              if (commandResult.code === 0 || commandResult.code === null) {
+                window.showErrorMessage(
```
Comment (chrjorgensen) [inline src/languages/clle/clRunner.ts:34]:
```suggestion
                window.showInformationMessage(
```

because I get this:

image

**Review by chrjorgensen — CHANGES_REQUESTED** [review]
@worksofliam Looking good - only a few minor issues...

File: package.json [hunk 3]
```diff
@@ -1629,6 +1640,10 @@
 				{
 					"command": "code-for-ibmi.sortMembersByName",
 					"when": "never"
+				},
+				{
+					"command": "code-for-ibmi.cl.runSelected",
+					"when": "code-for-ibmi:connected"
```
Comment (sebjulliand) [inline package.json:1646]: Only shows in palette if displaying a cl
```suggestion
					"when": "code-for-ibmi:connected && editorLangId === cl"
```

File: package.json [hunk 4]
```diff
@@ -1471,6 +1476,12 @@
 				"command": "code-for-ibmi.connectTo",
 				"key": "ctrl+alt+c",
 				"mac": "cmd+alt+c"
+			},
+			{
+				"command": "code-for-ibmi.cl.runSelected",
+        "key": "ctrl+r",
+        "mac": "cmd+r",
```
Comment (sebjulliand) [inline package.json:1483]: Keep it aligned 😅
```suggestion
                "key": "ctrl+r",
                "mac": "cmd+r",
```

File: src/languages/clle/clApi.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            window.showInformationMessage(`Failed to install CL components.`);
+          }
+          break;
+        }
+      });
+  }
+}
+
+function checkRequirements() {
+  /** @type {IBMi} */
```
Comment (sebjulliand) [inline src/languages/clle/clApi.ts:28]: This should be removed

File: src/languages/clle/clApi.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+async function install() {
+  const connection = instance.getConnection()!;
+  const content = instance.getContent()!;
+  const config = instance.getConfig()!;
+
+  const tempLib = config.tempLibrary;
+
+  try {
+    await connection.remoteCommand(`CRTSRCPF ${tempLib}/QTOOLS`, undefined)
```
Comment (sebjulliand) [inline src/languages/clle/clApi.ts:42]: While you're at it, why not use `runCommand`?

File: src/languages/clle/clApi.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  } catch (e) {
+    //It may exist already so we just ignore the error
+  }
+
+  await content.uploadMemberContent(undefined, tempLib, `QTOOLS`, `GENCMDXML`, gencmdxml.content.join(`\n`));
+  await connection.remoteCommand(
```
Comment (sebjulliand) [inline src/languages/clle/clApi.ts:48]: Same as above; use `runCommand`?

**Review by sebjulliand — CHANGES_REQUESTED** [review]
Some minor changes; overall a nice feature that will surely come in handy!
Nice work!

File: src/languages/clle/clApi.ts (hunk 6 above)
Comment (worksofliam) [reply src/languages/clle/clApi.ts:42]: I will do this in a second PR.

**Comment by worksofliam** [comment]
The issues have been resolved. Another review whenever you have a moment would be fab. Thanks!

**Review by sebjulliand — CHANGES_REQUESTED** [review]
A small issue needs to be fixed; otherwise it's brilliant! 😊

**Review by sebjulliand — APPROVED** [review]
Looking good, works as expected! Let's merge!

**Pull request merged**